### PR TITLE
Changes to inlineProjections to handle the query from #3828

### PR DESCRIPTION
**Pull request by systay**
Limits inlining if an expression is used too many times

### CODE REVIEW

**Comment by davidegrohmann** [comment]
@systay this guy needs a rebase

**Comment by systay** [comment]
Doing le rebase

File: community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/InlineProjectionsTest.scala [hunk 1]
```diff
@@ -382,6 +382,24 @@ class InlineProjectionsTest extends CypherFunSuite with AstRewritingTestSupport
       """.stripMargin))
   }
 
+  test("should not inline expressions used many times: WITH 1 as a MATCH (a) RETURN a => WITH 1 as a MATCH (a) RETURN a AS a") {
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/InlineProjectionsTest.scala]: test name is wrong, it uses a different query

File: community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/InlineProjectionsTest.scala [hunk 2]
```diff
@@ -382,6 +382,24 @@ class InlineProjectionsTest extends CypherFunSuite with AstRewritingTestSupport
...
+        |MATCH (a) WHERE a.prop = x OR a.bar > x
+        |RETURN a, x""".stripMargin)
+
+    result should equal(ast(
+      """WITH 1 as x
+        |MATCH (a) WHERE a.prop = x OR a.bar > x
+        |RETURN a, x""".stripMargin))
+  }
+
+  test("should not inline relationship identifiers if not inlinging expressions") {
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/InlineProjectionsTest.scala]: typo: inligning -> inlining

Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/InlineProjectionsTest.scala]: btw I don't understand the test, what kind of expressions are you referring to?

File: community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/InlineProjectionsTest.scala [hunk 3]
```diff
@@ -170,13 +170,13 @@ class InlineProjectionsTest extends CypherFunSuite with AstRewritingTestSupport
 
   test("should not inline identifiers into patterns: WITH 1 as a MATCH (a) RETURN a => WITH 1 as a MATCH (a) RETURN a AS a") {
     val result = projectionInlinedAst(
-      """WITH 1 as a
+      """WITH {node} as a
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/InlineProjectionsTest.scala]: you should also change the test name

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/InliningContext.scala [hunk 4]
```diff
@@ -47,26 +53,37 @@ case class InliningContext(projections: Map[Identifier, Expression] = Map.empty,
...
+  def isAliasedIdentifier(identifier: Identifier) = alias(identifier).nonEmpty
+
   def alias(identifier: Identifier): Option[Identifier] = projections.get(identifier) match {
     case Some(other: Identifier) => Some(other.copyId)
     case _                       => None
   }
 }
 
+object InliningContext {
+  val INLINING_THRESHOLD = 3
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/InliningContext.scala]: private?

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/InliningContext.scala [hunk 5]
```diff
@@ -47,26 +53,37 @@ case class InliningContext(projections: Map[Identifier, Expression] = Map.empty,
...
         case alias @ Some(_) => node.copy(identifier = alias)(node.position)
         case _               => node
       }
-    case rel @ RelationshipPattern(Some(ident), _, _, _, _, _) =>
+    case rel @ RelationshipPattern(Some(ident), _, _, _, _, _) if okToRewrite(ident) =>
       alias(ident) match {
-        case alias @ Some(_) => rel.copy(identifier = alias)(rel.position)
+        case alias @ Some(_) =>
+          val x = rel.copy(identifier = alias)(rel.position)
+          x
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/rewriters/InliningContext.scala]: revert

**Comment by davidegrohmann** [comment]
It looks fine, only some small details to fix.

**Pull request merged**